lib/compiler/include/compiler/unity_algorithm.h

```diff
@@ -1,24 +1,77 @@
 #ifndef _FLEXFLOW_FFC_UNITY_ALGORITHM_H
 #define _FLEXFLOW_FFC_UNITY_ALGORITHM_H
 
-#include "utils/graph.h"
-#include "op-meta/operator_attrs.h"
+#include "op-attrs/operator_attrs.h"
 #include "pcg/machine_view.h"
+#include "pcg/machine_specification.h"
+#include "utils/graph.h"
+#include "compiler.h"
+#include "cost_estimate.h"
 
 namespace FlexFlow {
 
-/* std::unordered_map<MultiDiEdge, ParallelTensorShape> infer_tensor_shapes(ParallelComputationGraph const &); */
+using OptimizerComputationGraph =
+    NodeLabelledMultiDiGraph<ComputationGraphAttrs>;
+using OptimizerPCG =
+    LabelledMultiDiGraph<PCGOperatorAttrs, ParallelTensorShape>;
+
+using SubParallelComputationGraph =
+    LabelledOpenMultiDiGraph<PCGOperatorAttrs,
+                             ParallelTensorShape,
+                             MachineView>;
+
+struct Substitution {
+};
+
+struct Strategy {
+  Strategy(float runtime, std::unordered_map<Node, MachineView> machine_views);
+  bool operator<(Strategy const &s) const;
+
+  static Strategy sequential_combine(Strategy const &s1, Strategy const &s2);
+  static Strategy parallel_combine(Strategy const &s1, Strategy const &s2);
+  static Strategy infinity();
+
+  float runtime;
+  std::unordered_map<Node, MachineView> machine_views;
+};
+
+Strategy
+    optimal_cost(OptimizerPCG const &g,
+                 std::function<std::unordered_set<MachineView>(
+                     PCGOperatorAttrs const &, MachineSpecification const &)> const
+                     &allowed_machine_views,
+                 ICostEstimator const &cost_estimator,
+                 MachineSpecification const &resources,
+                 std::unordered_map<size_t, Strategy> &cached_subgraph_costs);
+
+struct GraphOptResult {
+  OptimizerPCG pcg;
+  Strategy strategy;
+
+  GraphOptResult(OptimizerPCG const &pcg, Strategy const &strategy);
+
+  bool operator<(GraphOptResult const &r) const;
+};
+
+GraphOptResult
+    graph_optimize(OptimizerComputationGraph &cg,
+                   ICostEstimator const &cost_estimator,
+                   MachineSpecification const &resources,
+                   std::function<std::unordered_set<MachineView>(
+                       PCGOperatorAttrs const &, MachineSpecification const &)> const
+                       &allowed_machine_views,
+                   OptimizerConfig const &opt_config);
+                   
+} // namespace FlexFlow
 
-/* std::unordered_set<Node> get_nodes(Serial const &serial); */
-/* std::unordered_set<Node> get_nodes(Parallel const &parallel); */
-/* std::unordered_set<Node> get_nodes(Node const &node); */
-/* std::unordered_set<Node> get_nodes(SerialParallelDecomposition const &sp); */
+namespace std {
+  
+template <>
```

Comment:
Where is the hash implementation for this--I didn't see it in `unity_algorithm.cc`